### PR TITLE
add self.head_dim for VisionAttention in Qwen2-VL

### PR DESCRIPTION
**Pull request by GeLee-Q**
## Add self.head_dim for VisionAttention in Qwen2-VL

This PR adds the `self.head_dim` attribute to the VisionAttention class in the Qwen2-VL model. This addition is necessary for proper dimension calculations in the attention mechanism of the vision component.

### Changes made

- Added `self.head_dim` attribute to the VisionAttention class
- Initialized `self.head_dim` with the appropriate value

### Motivation

The `head_dim` attribute is crucial for calculating attention scores and outputs correctly. Its addition ensures that the vision attention mechanism in Qwen2-VL operates as intended, maintaining consistency with the model's architecture.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ✅] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ✅]] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@simonJJJ @zucchini-nlp @ArthurZucker

### CODE REVIEW

File: src/transformers/models/qwen2_vl/modeling_qwen2_vl.py [hunk 1]
```diff
@@ -275,6 +275,7 @@ class VisionAttention(nn.Module):
     def __init__(self, dim: int, num_heads: int = 16) -> None:
         super().__init__()
         self.num_heads = num_heads
+        self.head_dim = dim // num_heads
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_vl/modeling_qwen2_vl.py:278]: good, I am a bit  baffled as to how this was not caught, the math.sqrt could not have run 😅

Comment (ArthurZucker) [reply src/transformers/models/qwen2_vl/modeling_qwen2_vl.py:278]: We have few failing tests: https://github.com/huggingface/transformers/actions/runs/10656977518/job/29536379001#step:13:694 but this was not caught. 
```python
     @require_bitsandbytes
    def test_small_model_integration_test_batch_different_resolutions(self):
        model = Qwen2VLForConditionalGeneration.from_pretrained("Qwen/Qwen2-VL-7B-Instruct", load_in_4bit=True)
>       text, vision_infos = self.processor.apply_chat_template(
            self.messages, tokenize=False, add_generation_prompt=True
        )
E       ValueError: too many values to unpack (expected 2)
```
this one needs to be updated

Comment (GeLee-Q) [reply src/transformers/models/qwen2_vl/modeling_qwen2_vl.py:278]: Hello, I found the code related to vision_infos in the file vision_process.py on [QwenLM](https://github.com/QwenLM/Qwen2-VL/blob/main/qwen-vl-utils/src/qwen_vl_utils/vision_process.py). However, the Qwen2-VL processor in tranformers does not have an interface to process vision_info. Therefore, I added a function to process this information.

File: tests/models/qwen2_vl/test_modeling_qwen2_vl.py [hunk 2]
```diff
@@ -428,11 +429,16 @@ def test_small_model_integration_test_batch_different_resolutions(self):
...
             }
         ]
-        text2, vision_infos2 = self.processor.apply_chat_template(
-            messages2, tokenize=False, add_generation_prompt=True
-        )
+        text2 = self.processor.apply_chat_template(messages2, tokenize=False, add_generation_prompt=True)
+
+        vision_infos = self.extract_vision_info(messages2)
+        image_url = vision_infos[0]["image"]
+        image_input2 = Image.open(requests.get(image_url, stream=True).raw)
```
Comment (ArthurZucker) [inline tests/models/qwen2_vl/test_modeling_qwen2_vl.py:436]: I think that the processor is supposed to be able to handle urls or images and properly open them, would make sense to add this if it's not currently the case for an eased usage no? 🤗

**Review by ArthurZucker — APPROVED** [review]
Thanks, were you able to run all the tests for this? 🤗

**Comment by GeLee-Q** [comment]
> Thanks, were you able to run all the tests for this? 🤗

By commenting out @slow and @require_bitsandbytes, I ran test_small_model_integration_test_batch_wo_image locally. However, since I used the qwen2-vl-2b model, it caused an OOM error on an A800 single-card machine. I can ensure the code's correctness before `output = model.generate(**inputs, max_new_tokens=30)`. Tomorrow, I will test again using Qwen2-VL-2B-Instruct-GPTQ-Int4.

**Comment by GeLee-Q** [comment]
> Thanks, were you able to run all the tests for this? 🤗

@ArthurZucker  Hi! I ran these tests and encountered the following issues. Some of the test results show only minor deviations, which I believe are not significant issues. Two tests encountered OOM (Out of Memory) problems, and there are already existing issues in the model author's relevant repository.

#### Qwen2VLModelTest.test_batching_equivalence
```
transformers/tests/test_modeling_common.py:735: in recursive_check
    self.assertTrue(
E   AssertionError: tensor(False, device='cuda:0') is not true : Batched and Single row outputs are not equal in Qwen2VLForConditionalGeneration 
for key=logits. Difference=0.0031325221061706543.
```

#### Qwen2VLIntegrationTest.test_small_model_integration_test

```
// for pic
assert torch.allclose(expected_pixel_slice, inputs.pixel_values[:6, :3], atol=1e-3)
E       assert False
E        +  where False = <built-in method allclose of type object at 0x7f766a7242e0>(tensor([[0.8501, 0.8647, 0.8647],\n        [1.0106, 1.0106, 1.0252],\n        [0.9960, 1.0106, 1.0252],\n        [1.0982, 1.1128, 1.1274],\n        [1.0836, 1.0982, 1.0982],\n        [1.1858, 1.1858, 1.1858]]), tensor([[0.8501, 0.8501, 0.8647],\n        [0.9376, 0.9376, 0.9376],\n        [0.9084, 0.9376, 0.9376],\n        [1.0252, 1.0252, 1.0544],\n        [1.0252, 1.0252, 1.0252],\n        [1.0836, 1.0836, 1.0836]]), atol=0.001)
E        +    where <built-in method allclose of type object at 0x7f766a7242e0> = torch.allclose

// for llm output
E       AssertionError: 'syst[60 chars]this?\nassistant\nThe dog in the picture appea[117 chars]ices' != 'syst[60 chars]this?assistant\nThe dog in the picture appears[106 chars]ure,'
E       Diff is 686 characters long. Set self.maxDiff to None to see it.
```
#### Qwen2VLIntegrationTest.test_small_model_integration_test_batch_wo_image

> EXPECTED_DECODED_TEXT = [
>             "system\nYou are a helpful assistant.\nuser\nWhat kind of dog is this?assistant\nThe dog in the picture appears to be a Labrador Retriever. Labradors are known for their friendly and outgoing personalities, as well as their",
>             "system\nYou are a helpful assistant.user\nWho are you?assistant\nI am Qwen, a large language model created by Alibaba Cloud. I am designed to assist with various tasks and answer a wide range of questions to",
>         ]
> output_text :['system\nYou are a helpful assistant.\nuser\nWhat kind of dog is this?\nassistant\nThe dog in the picture appears to be a Labrador Retriever. Labradors are known for their friendly and intelligent nature, making them popular choices',
> 'system\nYou are a helpful assistant.\nuser\nWho are you?\nassistant\nI am a large language model created by Alibaba Cloud. I am called Qwen.']   

        
```
E       AssertionError: Lists differ: ['sys[61 chars]this?\nassistant\nThe dog in the picture appea[262 chars]en.'] != ['sys[61 chars]this?assistant\nThe dog in the picture appears[320 chars] to']
E       
E       First differing element 0:
E       'syst[60 chars]this?\nassistant\nThe dog in the picture appea[117 chars]ices'
E       'syst[60 chars]this?assistant\nThe dog in the picture appears[108 chars]heir'
E       
E       Diff is 1060 characters long. Set self.maxDiff to None to see it.
```

#### Qwen2VLIntegrationTest.test_small_model_integration_test_batch_different_resolutions
#### Qwen2VLIntegrationTest.test_small_model_integration_test_batch
These two tests will encounter insufficient video memory, but my environment is currently unable to successfully run GPTQ-int4 and GPTQ-int8 models. The related out of memory issues have been reported here [https://github.com/QwenLM/Qwen2-VL/issues/88](url)

**Comment by GeLee-Q** [comment]
@ArthurZucker @zucchini-nlp Hello, I was wondering if this PR could be merged. Based on this code, I discovered a  [precision issue with fp16 inference]( https://github.com/huggingface/transformers/issues/33294). I've temporarily resolved this issue by modifying the source code, and I will create a separate PR to address this problem.

**Comment by GeLee-Q** [comment]
> Hi!
> 
> Could you push an empty commit with message
> 
> `[run-slow] qwen2_vl`
> 
> Thanks!

Okay, I've submitted it. Is there anything else that needs to be done?

**Comment by ydshieh** [comment]
> Okay, I've submitted it. Is there any

Not at this moment :-) Have to wait CI's results.

**Comment by GeLee-Q** [comment]
> > Okay, I've submitted it. Is there any
> 
> Not at this moment :-) Have to wait CI's results.

@ydshieh hello, During CI, I encountered an error related to Python versioning. The list[dict] syntax is only supported in Python 3.9+.  This code originates from the [Qwen team's source code](https://github.com/QwenLM/Qwen2-VL/blob/main/qwen-vl-utils/src/qwen_vl_utils/vision_process.py), do you think this needs to be modified?

```
==================================== ERRORS ====================================
_______ ERROR collecting tests/models/qwen2_vl/test_modeling_qwen2_vl.py _______
tests/models/qwen2_vl/test_modeling_qwen2_vl.py:[30](https://github.com/huggingface/transformers/actions/runs/10702431916/job/29671213250?pr=33211#step:11:31)4: in <module>
    class Qwen2VLIntegrationTest(unittest.TestCase):
tests/models/qwen2_vl/test_modeling_qwen2_vl.py:459: in Qwen2VLIntegrationTest
    def extract_vision_info(self, conversations: list[dict] | list[list[dict]]) -> list[dict]:
E   TypeError: 'type' object is not subscriptable
=========================== short test summary info ============================
ERROR tests/models/qwen2_vl/test_modeling_qwen2_vl.py - TypeError: 'type' object is not subscriptable
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 2.28s ===============================
```

**Comment by ydshieh** [comment]
Great we catch it here!

Yes, we have to modify it as we are still running with python 3.8. (Will change in 2 months).

Could you try something like

> List[Dict]

> List[List[Dict]]

where they are imported like

> from typing import Dict, List

(you can search the codebase to see some such usages)

**Comment by ydshieh** [comment]
You probably also needs `Union`

**Comment by GeLee-Q** [comment]
> You probably also needs `Union`

@ydshieh  Thank you, I have adapted the code to be compatible with Python 3.8.

**Comment by GeLee-Q** [comment]
> Quote reply

@ydshieh Hi，could you please start the Action when it's convenient for you?

**Comment by GeLee-Q** [comment]
> You probably also needs `Union`

 Hi，could you please start the Action when it's convenient for you?

**Comment by GeLee-Q** [comment]
> > Okay, I've submitted it. Is there any
> 
> Not at this moment :-) Have to wait CI's results.

@ydshieh  Hi, The CI results are out. What else needs to be done to merge the code?  Do I need to submit an empty commit with the message "[run-slow] qwen2_vl" again?

**Comment by ydshieh** [comment]
oh, yes. You need a commit of that message.

as you can see, so far it is

> PR slow CI / Run all tests for the model (pull_request) Skipped

**Comment by GeLee-Q** [comment]
> oh, yes. You need a commit of that message.
> 
> as you can see, so far it is
> 
> > PR slow CI / Run all tests for the model (pull_request) Skipped

Okay, thanks

**Comment by zucchini-nlp** [comment]
We'll merge https://github.com/huggingface/transformers/pull/33161 soon, which should add more tests and fix slow CI. You can then rebase `main` to get a green CI :)

**Comment by GeLee-Q** [comment]
> We'll merge #33161 soon, which should add more tests and fix slow CI. You can then rebase `main` to get a green CI :)

Okay, thank you. I've seen their modifications and commits. The test code has resolved the precision and OOM (Out of Memory) issues. My modifications to the test code should be unnecessary now.

**Comment by GeLee-Q** [comment]
> We'll merge #33161 soon, which should add more tests and fix slow CI. You can then rebase `main` to get a green CI :)

@zucchini-nlp Hello, I've modified and resolved the conflicting code. Could you please trigger the action and merge it when convenient? Thank you!

**Review by zucchini-nlp — APPROVED** [review]
Thanks for fixing! Will merge shortly, can you rebase main please

**Comment by zucchini-nlp** [comment]
Yep, sure. @GeLee-Q can you tag me when you're done rebasing/testing and add the last commit with `[run-slow] qwen2_vl` message, I'll approve slow CI run

**Comment by GeLee-Q** [comment]
@zucchini-nlp Hi !  I've completed the rebase and pushed the [run-slow] commit for qwen2_vl. The branch is ready for your review and approval for the slow CI run. Let me know if you need anything else.

**Comment by zucchini-nlp** [comment]
Tests for sdpa are failing on multi-gpu setting but from the logs seems like the diff is around `1e-03`. The error doesn't seem to be caused by this PR and is passing on single-gpu. I think we can merge, what do you say @ydshieh?

**Comment by ydshieh** [comment]
I am running on `1759bb9126e59405f58693a17ef9f58040c2008b` (main) which is the base of this PR.

The test is passing there.

https://github.com/huggingface/transformers/actions/runs/10737126379/job/29777960925

Not sure if it's flaky. We can re-trigger CI  here.

**Comment by zucchini-nlp** [comment]
@ydshieh yes, apparently it is passing and the CI is green now. Do you want me to retrigger CI again?

**Pull request merged**